### PR TITLE
Autodeploy docs on release

### LINKED ISSUE #1: Move docs into gh-pages branch
Command reference is currently generated with Sphinx locally, and pushed to the repository alongside code. This is problematic because..

1. The user need all dependencies (Maya, Sphinx, nose, plug-ins) in order to update the docs
2. Repository bloated with generated content, hard to know what to edit when making changes.

Instead, each push should trigger a re-build of the documentation and upload into the gh-pages branch of this repository.

### PR DESCRIPTION
**Pull request by monkeez**
Resolves #1 

- Added pipeline step for deploying generated Sphinx documentation to `gh-pages` branch.
- Removed generated documentation from development branch.
- Added `gh-pages` branch to the exclusion list for triggering pipeline runs.

### Walkthrough of the changes
After all the steps in the `Maya` job I added an extra one to save the generated docs as an artifact:
```yaml
- publish: $(System.DefaultWorkingDirectory)/build/html
  artifact: docs
  condition: and(succeeded(), eq(variables['containerImage'], 'mottosso/maya:2020'))
```
This is only saving the docs from the Maya 2020 run, as we only need one copy.

Next I added the `Deploy_Docs` job which handles the publishing of documentation only on releases. This has 4 steps:
- Download the saved artifact
- Initialise a new git repo
- Install an SSH key from Azure Pipelines and add Github as a known host
- Push this repo to the `gh-pages` branch on Github

This was the simplest way of achieving this that I could find after doing some research, mostly sourced from [here](https://github.com/microsoft/CCF/blob/master/.azure-pipelines-gh-pages.yml) but also other locations.

### Changes to make on your end
Firstly you'll need to create an orphaned `gh-pages` branch, remove all files, commit, and then push it to Github.
```
git checkout --orphan gh-pages
git rm -rf .
rm '.gitignore'
git commit -m 'Removed files'
git push origin gh-pages
```
Then change Github pages to use the `gh-pages` branch in the repository settings.

Finally you'll have to generate an SSH key for Github and Azure Pipelines to use for deploying, and add a secure environment variable for Github as a known host. I followed along with [this](https://docs.microsoft.com/en-us/azure/devops/pipelines/tasks/utility/install-ssh-key?view=azure-devops#example-setup-using-github) to set it up on my end. The key should be called `deploy_key` and the environment variable should be `KNOWN_HOST` to work with my current script, but you're more than welcome to change these in the `azure-pipelines.yaml`.

### Potential issues
- Currently I've got the git `user.email` set to my own email, I wasn't sure what to put here, if you wanted your email or another one.
- This technique doesn't save any commit history for the generated docs as it's initialising a new repo to push from each build. I figured this should be fine as it's only generated content and the history of the actual changes to the source files will still be present in the `master` branch.

If you have any questions please ask!

### CODE REVIEW

**Comment by mottosso** [comment]
Holy! Thanks for this, this looks like a massive headache that I'm glad not to have again xD

> Currently I've got the git user.email set to my own email

These shouldn't get uploaded, I think these are local to your machine (either per repo, or per user account)

> This technique doesn't save any commit history for the generated docs as it's initialising a new repo to push from each build

That's fine and expected.

Running through your list of changes on my end now. I suspect we can't actually make sure this works, without actually merging and making a release. That's fine, we can make a few test releases and delete them as we go. Part of the headache still lies ahead xD

**Comment by monkeez** [comment]
Yeah, I did some tests on my fork where I didn't have the condition of only deploying on a release. It worked out OK after some fiddling https://monkeez.github.io/cmdx/

But yes, to truly test it I think we'll have to do some releases :D.

**Comment by mottosso** [comment]
Meeeeh, headaches are settling in. xD You should be able to push that orphan branch yourself I think? You are welcome to, don't sweat it. I'm on a new machine and need to muck about with permissions and SSH before this can happen.

The Azure stuff already has some credentials in it, for releasing to PyPI, maybe we could leverage that? Otherwise I'll set something up.

**Comment by monkeez** [comment]
I've pushed the `gh-pages` branch.

I'm not sure the credentials for PyPI can be used for this. I would set it all up myself but I inderstand that's not really possible.

File: azure-pipelines.yml [hunk 1]
```diff
@@ -67,13 +68,60 @@ jobs:
...
+    steps:
+      - task: DownloadPipelineArtifact@2
+        inputs:
+          artifact: docs
+          path: $(Build.SourcesDirectory)/build/html
+
+      - script: |
+          git init
+          git config --local user.name "Azure Pipelines"
+          git config --local user.email "mitchellcoote@gmail.com"
```
Comment (mottosso) [inline azure-pipelines.yml:103]: Ah I see this is what you meant with `user.name`. This is fine, let it be known who accomplished this feat. :)

**Comment by mottosso** [comment]
Ok, that should be it. I'm not sure what GitHub is on about, the build has run successfully on Azure. Let's merge, and please go ahead and trigger a few releases to test things out. You can delete releases too, the UI on GitHub can be a little tricky to do that (probably for good reason). Let me know if you run into any trouble.

**Comment by monkeez** [comment]
Hmm, something went wrong with the install SSH key task.

**Comment by mottosso** [comment]
Having a look. Does it need the private or public key?

**Comment by mottosso** [comment]
Private key is up, here's what I've got but it still isn't working. :/

image

**Comment by mottosso** [comment]
Bah. I don't know what it wants. I'm out of time to play with this today. Happy to grant you access to the Azure site, next time I'm on this. If you have any suggestions, that would be good too.

**Comment by monkeez** [comment]
Yeah that looks right, the private key on Azure. Did you add the public part of the key to [Github Deploy keys](https://github.com/mottosso/cmdx/settings/keys) with write access?

**Comment by mottosso** [comment]
Yep

**Comment by monkeez** [comment]
Would it be possible to get access to the GitHub repo and Azure Pipeline settings?

**Comment by mottosso** [comment]
I wasn't able to enable that on GitHub, looks like only organisation accounts can do that. But you should be able to edit the Azure settings now, have a look.

**Comment by monkeez** [comment]
I've added my private key that worked for me to the secure files, you'll have to rename it to `deploy_key` as it won't let me delete the old one. Is there any way I can send you the public part of the key so you can update the GitHub settings?

**Comment by mottosso** [comment]
I've deleted the old one, and renamed yours to deploy_key. Send your public key to me@mottosso.com and I'll add it.

**Comment by monkeez** [comment]
Sent 🤞

**Comment by mottosso** [comment]
I think we did it! Thanks for jumping through all of these hoops @monkeez. :) Now the docs will be in sync with the code, which is fabulous!

**Pull request merged**